### DISCUSSION TITLE
How to implement a gdb-like watch command

### DISCUSSION THREAD

**Discussion by wangziming133** [bpftrace/bpftrace#2374] [Q&A]
How does watchpoint print out the address?
```
bpftrace -e 'watchpoint:0x123456:4:w { printf("address"); exit(); }'
```

in gdb
> (gdb) attach pid
> (gdb) watch *0x123456
 watchpoint 1: *0x123456
> (gdb) continue
0x00001 in ??

**Comment by viktormalik** [comment]
I'm not sure what you try to achieve. If you want to print the watchpoint address when it is hit (`0x123456`), you can use the `probe` builtin.

**Reply by wangziming133** [reply]
I want to get the blue address below.

![image](https://user-images.githubusercontent.com/108342252/195289870-37df8ad5-cfdd-4ffb-ab4a-3b908cdfd5d2.png)

**Reply by ajor** [reply]
That's the address of the instruction that's doing the write, which you can access through the `ip` register on x86:
`watchpoint:0x123456:4:w { printf("%p\n", reg("ip")) }`

**Reply by iDings** [reply]
@ajor Any method to fetch the odl/new value like as the gdb output?

**Reply by ajor** [reply]
Getting the new value is easy - just read the data from your watchpoint address. For the old value, I think you'll have to have your script remember it by itself.

It's a bit more manual that using a debugger, but AFAIK this is the same approach they take under the hood:
```
uprobe:./a.out:main
{
  // initialise old_val at program start
  @old_val = *0x40401c;
}
watchpoint:0x40401c:4:w
{
  $new_val = *0x40401c;
  printf("Old val = %d\nNew val = %d\n", @old_val, $new_val);
  @old_val = $new_val;
}
```

**Reply by iDings** [reply]
Got it, thanks.

**Reply by iDings** [reply]
There are `unwatch` function to remove watchpoint,  but no `watch` function, any possible to add `watch` function to create watchpoint dynamically? Because some memory or struct object is allocated dynamically, if we can add watchpoint dynamically, then can observe the write/read of a member in object.

**Reply by ajor** [reply]
There's nothing that'd technically stop us from adding a `watch` function, but it'd need a good syntax (where is the probe's body defined?) and might lead to questions about attaching all probe types dynamically (again theoretically possible).

Dynamic attaching would have to happen asynchronously, so you might miss some updates between calling `watch` and the new watchpoint actually being enabled.

**Reply by iDings** [reply]
I did some analysis about the ebpf flow, but still a half-bakedt and immature idea.

> but it'd need a good syntax (where is the probe's body defined?)

Need a unique name to link watch function with the probe body, In internal every probe body seems will generate an individual bpf prog, bpftrace need record the map that combine unique name with bpf prog context, and use it  in generated `watch` bpf code to bind watchpoint to the desired probe bpf prog.

```
tracepoint:...
{
  watch(addr, 4, r, onwatch_func)
}

watchpoint:onwatch_func
{
}
```

> Dynamic attaching would have to happen asynchronously, so you might miss some updates between calling watch and the new watchpoint actually being enabled.

Dynamic attaching watchpoint asynchronously seems not the ideal way,  need add a new bpf function(`struct bpf_func_proto`) in kernel to make it synchronize, and watchpoint should be added on any kernel context(irq, softirq).
